test/type_traits.cpp

```diff
@@ -530,6 +537,215 @@ TEST_CASE("is_convertible", "[meta.rel]") {
+    CHECK_NESTED(not_invocable<Fnl, C const&, int>());
+    CHECK_NESTED(not_invocable<Fnl, C&&, int>());
+    CHECK_NESTED(not_invocable<Fnl, C const&&, int>());
+
+    using Fnr = int (C::*)(int)&&;
+
+    CHECK_NESTED(not_invocable<Fnr, C&, int>());
+    CHECK_NESTED(not_invocable<Fnr, C const&, int>());
+    CHECK_NESTED(invocable<int, Fnr, C&&, int>(throws));
+    CHECK_NESTED(not_invocable<Fnr, C const&&, int>());
+
+    using Fncl = int (C::*)(int) const&;
+
+    CHECK_NESTED(invocable<int, Fncl, C&, int>(throws));
+    CHECK_NESTED(invocable<int, Fncl, C const&, int>(throws));
+    CHECK_NESTED(invocable<int, Fncl, C&&, int>(throws));
+    CHECK_NESTED(invocable<int, Fncl, C const&&, int>(throws));
+
+    using Fncr = int (C::*)(int) const&&;
+
+    CHECK_NESTED(not_invocable<Fncr, C&, int>());
+    CHECK_NESTED(not_invocable<Fncr, C const&, int>());
+    CHECK_NESTED(invocable<int, Fncr, C&&, int>(throws));
+    CHECK_NESTED(invocable<int, Fncr, C const&&, int>(throws));
+  }
+
+  /* obj-mem-ptr */ {
+    using Fn = int C::*;
+
+    CHECK_NESTED(invocable<int&, Fn, C&>(nothrows));
+    CHECK_NESTED(invocable<int const&, Fn, C const&>(nothrows));
+    CHECK_NESTED(invocable<int&&, Fn, C&&>(nothrows));
+    CHECK_NESTED(invocable<int const&&, Fn, C const&&>(nothrows));
+
+    CHECK_NESTED(invocable<int&, Fn, D&>(nothrows));
+    CHECK_NESTED(invocable<int const&, Fn, D const&>(nothrows));
+    CHECK_NESTED(invocable<int&&, Fn, D&&>(nothrows));
+    CHECK_NESTED(invocable<int const&&, Fn, D const&&>(nothrows));
+
+    CHECK_NESTED(invocable<int&, Fn, std::reference_wrapper<C>>(nothrows));
+    CHECK_NESTED(
+        invocable<int const&, Fn, std::reference_wrapper<C const>>(nothrows));
+
+    CHECK_NESTED(invocable<int&, Fn, C*>(nothrows));
+    CHECK_NESTED(invocable<int const&, Fn, C const*>(nothrows));
+
+    CHECK_NESTED(invocable<int&, Fn, smart_ptr<C>>(nothrows));
+    CHECK_NESTED(invocable<int const&, Fn, smart_ptr<C const>>(nothrows));
+
+    CHECK_NESTED(invocable<int&, Fn, smart_ptr_throws<C>>(throws));
+    CHECK_NESTED(invocable<int const&, Fn, smart_ptr_throws<C const>>(throws));
+
+    CHECK_NESTED(not_invocable<Fn>());
+    CHECK_NESTED(not_invocable<Fn, int&>());
+    CHECK_NESTED(not_invocable<Fn, C&, int>());
+  }
+
+  /* fun-obj */ {
+    struct Fn {
+      int operator()(int) noexcept { return 42; }
+    };
+
+    CHECK_NESTED(invocable<int, Fn, int>(nothrows));
+    CHECK_NESTED(invocable<int, Fn, conv_to<int>>(nothrows));
+    CHECK_NESTED(invocable<int, Fn, conv_to_throws<int>>(throws));
+    CHECK_NESTED(invocable<conv_from<int>, Fn, int>(nothrows));
+    CHECK_NESTED(invocable<conv_from_throws<int>, Fn, int>(nothrows, throws));
+    CHECK_NESTED(invocable<conv_from_throws<int>, Fn, conv_to_throws<int>>(
+        throws, throws));
+
+    CHECK_NESTED(not_invocable<Fn>());
+    CHECK_NESTED(not_invocable<Fn, void*>());
+    CHECK_NESTED(not_invocable<Fn, int, int>());
+
+#if __cpp_noexcept_function_type
```

Comment:
Do we want to replace this with `p0012`?

Reply:
Yes, missed this one.